tidy3d/components/device/simulation.py

```diff
@@ -318,18 +356,80 @@ def check_medium_heat_spec(cls, values):
             ),
         )
 
-        failed = cls._check_cross_solids(sim_box, values)
+        failed_solid_idx, failed_elect_idx = cls._check_cross_solids(sim_box, values)
 
-        if len(failed) > 0:
-            raise SetupError(
-                "No solid materials ('SolidSpec') are detected in heat simulation. Solution domain is empty."
-            )
+        simulation_types = cls._check_simulation_types(values=values)
+
+        for sim_type in simulation_types:
+            if sim_type == DeviceSimulationType.HEAT:
+                if len(failed_solid_idx) > 0:
+                    raise SetupError(
+                        "No solid materials ('SolidSpec') are detected in heat simulation. Solution domain is empty."
+                    )
+            elif sim_type == DeviceSimulationType.CONDUCTION:
+                if len(failed_elect_idx) > 0:
+                    raise SetupError(
+                        "No conducting materials ('ConductorSpec') are detected in conduction simulation. Solution domain is empty."
+                    )
+
+        return values
+
+    @staticmethod
+    def _check_simulation_types(
+        values: Dict,
+        HeatBCTypes=HeatBCTypes,
+        ElectricBCTypes=ElectricBCTypes,
+        HeatSourceTypes=HeatSourceTypes,
+    ) -> list[DeviceSimulationType]:
+        """Given model dictionary ``values``, check the type of simulations to be run
+        based on BCs and sources.
+        """
+        # NOTE: this function does not consider the source 'HeatFromElectricSource' by default
+
+        boundaries = values["boundary_spec"]
+        sources = values["sources"]
+
+        simulation_types = []
+
+        for boundary in boundaries:
+            if isinstance(boundary.condition, HeatBCTypes):
```

Comment:
could these be simplified  / more declarative by for example adding class attributes in some classes (eg. Boundary and Source) indicating which simulation types they belong to? Or is this circular?

Then this could be a bit more like

```py
simulation_types = []
for boundary in boundaries:
    simulation_types += boundary.condition.simulation_types
    ...
simulation_types = set(simulation_types)
```

In general, I think it's nice to define as much as we can as data in the classes, in an attempt to minimize the complexity of these operations. Basically offload the work to the data structure